Should resample automatically subtract background?

_Issue [JP-2024](https://jira.stsci.edu/browse/JP-2024) was created on JIRA by [James Davies [X]](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies):_

I believe the answer should be "no".  But currently `resample` does subtract off the sky background in `calwebb_image3` when it has background measurements available.  A scalar sky value gets computed earlier on in the stage3 pipeline via `skymatch`.

The single-resampled images produced at the end of the stage2 pipeline do not have the background subtracted, as they don't run through `skymatch`.

The current implementation checks if `skymatch` has computed a background level for each exposure.  If it has, it subtracts it before resampling.  So the output resampled images currently have a mean sky background of zero.

I suspect what it should do is subtract off those computed sky backgrounds for each exposure, drizzle into the common output image, and the add back the sky background after drizzling is completed, so that the output drizzled image has the same background as the input images.

But which sky background from which input image should be added back in?  The mean of the input sky backgrounds?  The minimum?  Some weighted mean?

I'm primarily concerned about what should be the default for SPD processing of the associations that we will be producing. I suspect all exposures in these associations will be taken close to the same time and should thus have pretty similar zodiacal light characteristics so pretty similar backgrounds.

_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2024?focusedCommentId=566949#comment-566949):_

Not to distract from Mihai's comments, but the requested changes result in differences to final products in the regression tests, e.g. test_nircam_image_stage3:

```python
AssertionError: 
   fitsdiff: 5.0
   a: /data1/jenkins/workspace/RT/JWST-Developers-Pull-Requests/clone/test_outputs/popen-gw14/test_nircam_image_run_detector1pipeline0/jw42424-o002_t001_nircam_clear-f444w_i2d.fits
   b: /data1/jenkins/workspace/RT/JWST-Developers-Pull-Requests/clone/test_outputs/popen-gw14/test_nircam_image_run_detector1pipeline0/truth/jw42424-o002_t001_nircam_clear-f444w_i2d.fits

Primary HDU:
  
     Headers contain differences:
       Keyword BKGLEVEL has different values:
          a> 0.0
          b> 0.1508732994059528
  
  Extension HDU 1 (SCI, 1):
  
     Data contains differences:
       Data differs at [1565, 4]:
          a> 0.15663332
          b> 0.005760026
       Data differs at [1566, 4]:
          a> 0.12281047
          b> -0.028062837
       Data differs at [1567, 4]:
[...]
  Extension HDU 8 (HDRTAB, 1):
  
     Data contains differences:
       Column BKGLEVEL data differs in row 0:
          a> 0.0
          b> 0.15087329940595284
       Column BKGLEVEL data differs in row 1:
          a> 0.0
          b> 0.15087329940595284
       Column BKGLEVEL data differs in row 2:
```

Is this the intended result, that the saved background values are now different?

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,10 @@
 1.4.2 (Unreleased)
 ==================
 
+skymatch
+--------
+
+- Changed default value of ``skymethod`` step parameter to 'match' [#6580]
 
 1.4.1 (2022-01-15)
 ==================

--- a/docs/jwst/skymatch/README.rst
+++ b/docs/jwst/skymatch/README.rst
@@ -90,7 +90,7 @@ The ``skymatch`` step has the following optional arguments:
 * ``skymethod`` (str):
   The sky computation algorithm to be used.
   Allowed values: {``local``, ``global``, ``match``, ``global+match``}
-  (Default = ``global+match``)
+  (Default = ``match``)
 
 * ``match_down`` (boolean):
   Specifies whether the sky *differences* should

--- a/jwst/skymatch/skymatch_step.py
+++ b/jwst/skymatch/skymatch_step.py
@@ -34,7 +34,7 @@ class SkyMatchStep(Step):
 
     spec = """
         # General sky matching parameters:
-        skymethod = option('local', 'global', 'match', 'global+match', default='global+match') # sky computation method
+        skymethod = option('local', 'global', 'match', 'global+match', default='match') # sky computation method
         match_down = boolean(default=True) # adjust sky to lowest measured value?
         subtract = boolean(default=False) # subtract computed sky from image data?
 

